app/models/user.rb

```diff
@@ -1,25 +1,63 @@
 class User < ApplicationRecord
   include Cul::Omniauth::Users
-  ROLES = %i[admin superadmin]
+
+  has_many :permissions, dependent: :destroy
+
   before_validation :add_email
 
   def password
     Devise.friendly_token[0,20]
   end
 
   def password=(*val)
-
   end
 
   def admin?
-    self.role == "admin"
+    can? :manage, :all
+  end
+
+  def editor?
+    !editable_libraries.count.zero?
   end
 
-  def superadmin?
-    self.role == "superadmin"
+  def editable_libraries
+    self.permissions
+      .where(action: 'edit', subject_class: Library.to_s)
+      .where.not(subject_id: nil)
+      .map{ |p| Library.find(p.subject_id) }
+  end
+
+  # Updating permissions. Destroys all previously definited permissions.
+  # Recreates them based on the paramters given. If admin flag is true,
+  # adds adnub access (can :manage, all). If not admin, recreates editor
+  # permissions using the array given.
+  def update_permissions(params)
+    admin = params.fetch(:admin, false)
+    library_ids = params.fetch(:library_ids, [])
+
+    admin = ActiveRecord::Type::Boolean.new.cast(admin) # Convert to bool.
```

Comment:
I need to cast it to a boolean because the parameter comes in from the user interface as 'true' or 'false' and it needs to be converted.